addons/website_sale_digital/models/ir_attachment.py

```diff
@@ -1,11 +1,37 @@
 # -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
-from odoo import fields, models
+from odoo import fields, models, api, _
+from odoo.exceptions import AccessError
 
 
 class Attachment(models.Model):
 
     _inherit = ['ir.attachment']
 
     product_downloadable = fields.Boolean("Downloadable from product portal", default=False)
+
+    @api.model
+    def check(self, mode, values=None):
+        def can_access(res_id, res_model):
+            # Mimick '/my/download' business code
+            purchased_products = self.env['account.invoice.line'].get_digital_purchases()
+
+            if res_model == 'product.product':
+                if res_id in purchased_products:
+                    return True
+
+            elif res_model == 'product.template':
+                template_ids = self.env['product.product'].sudo().browse(purchased_products).mapped('product_tmpl_id').ids
+                if res_id in template_ids:
+                    return True
+
+            return False
+
+        if self and not (self.env.user._is_admin() or self.env.user.has_group('sales_team.group_sale_salesman_all_leads')):
+            self._cr.execute('SELECT res_id, res_model, product_downloadable FROM ir_attachment WHERE id IN %s', [tuple(self.ids)])
+            for res_id, res_model, product_downloadable in self._cr.fetchall():
+                if product_downloadable and not can_access(res_id, res_model):
+                    raise AccessError(_("Sorry, you are not allowed to access this document."))
```

Comment:
I think we can do something lighter, considering that the controller uses `sudo()` to access the digital downloads. We really don't want to make this thing slower than it already is. And basically the only thing we need is to prevent `read` by non-employee users when `product_downloadable` is true.

Some pseudo-untested-code (maybe short-circuit UID 1 too?):
```py
    @api.model
    def check(self, mode, values=None):
        super().check(mode, values=values)
        if mode == 'read' and not self.env.user.has_group('base.group_user'):
            self._cr.execute('SELECT 1 FROM ir_attachment WHERE product_downloadable AND id IN %s',
                             [tuple(self.ids)]))
            if self._cr.rowcount():
                raise AccessError()
```


I would also not handle the case of employees, as that sounds out of scope for such a trivial module based on attachments. You'd need a proper table for tracking digital goods, with customizable access control, etc.

Reply:
> And basically the only thing we need is to prevent read by non-employee users when product_downloadable is true.

Only if the user hasn't purchased the product, no?

Reply:
The discussion with Romain yesterday was, you can download if one of
- admin (as always)
- have purchased the product
- have write access on the product
And obviously, relevant only for the `check` in `read` mode.

Reply:
> Only if the user hasn't purchased the product, no?

I would say: block it all the time. If you have purchased the product, the only place where you're supposed to access it is through the `/my/download?attachment_id=<id>` route, i.e. the direct link you receive after purchase. That controller already takes care of determining whether you are entitled to the download, then accesses the data in sudo mode, so there seems to be no need to duplicate that code..

> The discussion with Romain yesterday was, you can download if one of
> - admin (as always)
> - have purchased the product
> - have write access on the product

IMO, fine-grained access for employees is out of scope for such a trivial module, and it's kind of a new feature, too. Not sure it would be incredibly useful either. The main thing that we want here (in 12.0+) is to restrict direct access for portal users, in the least expensive manner. So outside of the specific case of portal users, I would just let the "normal" permissions apply.

Reply:
I do agree, I went too far and it was overkill.
As the business code is supposed to check you actually bought the product if you access it through normal workflow, we can prevent the read of digital product to non employee all the time.

Changes done.

But SQL query is skyrocketing, eg `/web/image` went from 13 to 17 queries, same for assets etc.

For master, wouldn't this work/be enough to not impact SQL query number?
```xml
<record id="website_digital_product_rule_portal" model="ir.rule">
    <field name="name">Digital product: only employee can read</field>
    <field name="model_id" ref="base.model_ir_attachment"/>
    <field name="domain_force">[('product_downloadable', '=', False)]</field>
    <field name="groups" eval="[(4, ref('base.group_portal')), (4, ref('base.group_public'))]"/>
    <field name="perm_read" eval="True"/>
    <field name="perm_write" eval="False"/>
    <field name="perm_create" eval="False"/>
    <field name="perm_unlink" eval="False"/>
</record>
```

Reply:
Here is what I suggest regarding my last message and what we just discussed with @mart-e 

Merge this PR up-to 14.0.
In the 14.0 forward-port, remove this code and add the `ir.rule`. This will also prevent the issue, without adding SQL queries, but won't work without a `-u`.

I think it is a good trade off, as it won't affect the perf improvements we did recently in 14.0, and most DB's will be fixed without a `-u` (12, 13, new db).